### ISSUE TITLE
Uncaught TypeError: Cannot delete property 'find a clean mug' of #<Object>

### ISSUE THREAD

**Issue by dschalk** [mweststrate/mobservable#36]
I have greatly improved the project running at http://transcendent.ninja. The README.md is out of date, but the code can be found at https://github.com/dschalk/mobservable-monads .
Today, I have been working on a monad implementation of "Todo List". The store is a reactive object instead of an array, so instead of "splice" I used "delete" followed by the index of the embedded object to be deleted. That results error messages such as: 
```
Uncaught TypeError: Cannot delete property 'find a clean mug' of #<Object>
```
Here is the constructor of the monad at issue:
```javascript
let M = new MonadObject(
   {
   'find a clean mug': {
       title: 'Find a clean mug',
       completed: true,
       remove: <button onClick = {() =>  {delete(M.x['find a clean mug']) }}
         style={{fontSize: 22 }}
         >
         Remove 
         </button>
    },

   'make coffee': {
       title: 'Make coffee',
       completed: false,
       remove: <button onClick = {() =>  {delete(M.x['Make coffee']) }}
         style={{fontSize: 22 }}
         >
         Remove
         </button>
    }
  }
);
```
Here is what appears in the console log:
```javascript	
var M = new MonadObject({
	  'find a clean mug': {
	    title: 'Find a clean mug',
	    completed: true,
	    remove: _react2['default'].createElement(
	      'button',
	      { onClick: function () {
	          delete M.x['find a clean mug'];
	        },
	        style: { fontSize: 22 }
	      },
	      'Remove'
	    )
	  },
```
Hovering the mouse pointer over the little "x" displays: "Uncaught TypeError: Cannot delete property 'find a clean mug' of #\<Object>".

What I am trying to do seems quite reasonable. I am allowed to splice elements out of arrays, so it is doubly surprising to find that I can't delete properties of an object. I wonder:

1. Is React or Mobservable causing this?

1. If it is Mobservable, did you intend to include use cases such as mine in the prohibition against deleting properties of objects.

2. Is there a work-around that can make it possible for me to remove embedded objects from objects.

My problem has wider implications. Shopping carts are real world applications that need "remove item" buttons. So if Mobservable is responsible for this error and there is a way to remove properties from objects, I suggest that the documentation make it easy to find so developers can quickly get on with their work.

**Comment by mweststrate** [comment]
1. It is caused by Mobservable, currently property delete's are not allowed on objects. 

2. So far objects are mostly considered to be structs, you want to observe the mutable properties individually in other observers. Deleting is to observers mostly the same as setting a property to `undefined`. However if you are looping the properties `delete` is quite different.

3. I think this is valid workaround (besides directly assigning `undefined`):
```javascript
var prop = Object.getOwnPropertyDescriptor(x, "Find a clean mug")
prop.enumerable = false;
Object.defineProperty(x, "Find a clean mug", prop);
x["Find a clean mug"] = undefined;
```

I think there are three possible solutions:
1. Allow deleting of properties. This would leak some memory though as Mobservable cannot detect this event (until ES7 is standardized) and will leak some memory
2. Provide a utiity method: `mobservable.deleteObservableProperty(x, "Find a clean mug")` which does all the clean up
3. Provide an `ObservableMap` classes with explict set, get, delete methods and an observable key collection to express that this object is a coherent set of key / values (just like an array is a coherent set of indexes / values) instead of an uncoherent set of preoperties.

**Comment by dschalk** [comment]
I have a version of Todo's at http://transcendent.ninja that works perfectly in Chrome and Opera in my Ubuntu 14.04 box, but not in Firefox. I uploaded the code to https://github.com/dschalk/mobservable-monads. While seeing what worked where, I discovered that the app at https://mweststrate.github.io/mobservable/getting-started.html doesn't work in any of my browsers.

The workaround you suggested didn't work, but that might have been due to my doing something wrong. I was just experimenting with storing todo's in an object. 

You probably have more interesting and pressing things to think about, but for the record, here's the code that works in Chrome and Opera: First, my simple starting todo's store:
```javascript
let M = new MonadArray(
   [
      [ 'find a clean mug', true ],
       [ 'make coffee', false ]
    ]
);
```
The display function is:
```javascript
displayTodos = () => {
  let ar = [];
  let M = this.M
  let l = this.M.x.length;
  for (let k=0;k<l;k+=1) {
    ar.push(  [ 
		<div>
			<div style={{fontSize: 22, color: 'orange' }} > {M.x[k][0]} </div>
			<div onClick={() => {M.fmap(this.removeTask, M.x[k][0]) }} 
				style={{color: 'lightblue', fontSize: 22 }} > Remove  </div>
			<div onClick={() => {M.fmap(this.toggleComplete, M.x[k][0]) }} 
				style={{color: 'lightblue', fontSize: 22 }} > 
				Complented:  <input type='checkbox' checked={M.x[k][1]} 
				  style={{width: 18, height: 18}} ></input>
			</div>
			<div onClick={() => {M.x[k][0] = prompt('Todo:', 'Take a walk') }} 
				style={{fontSize: 22, color: 'lightgreen'}}    
				  > 
				edit 
			</div>
			<br />
		</div> ]) }
    ar.push( <div style={{color: 'lawngreen' }}> {this.completed()} of {l} complete. </div> )
    ar.push( <div onClick={() => { let t = prompt('New task: ', 'smell the roses');
                                   M.fmap(this.addTask, t) }}
        style={{fontSize: 22, color: 'orange'}} 
      > 
    new todo 
  </div> )
      return ar;
}
```
Maybe Firefox would run this if I made it into a React component. Here are the other functions:
```javascript
    addTask = (mon,t) => {
      mon.x.push( [t,false])
      return mon;
    }

    removeTask = (mon,t) => {
      let l = mon.x.length;
      let k;
      for (k=0;k<l;k+=1) {
        let task = mon.x[k][0];
        if (mon.x[k][0] === t) {
          mon.x.splice(k, 1);
          console.log(mon.x);
          return mon;
        }
      }
      return mon;
    }

    toggleComplete = (mon,t) => {
      let l = mon.x.length;
      let k;
      for (k=0;k<l;k+=1) {
        if ( mon.x[k][0] === t) {
          mon.x[k][1] = !mon.x[k][1];
        }
      }
      return mon;
    }

    let completed = () => { return (M.x.filter(y => y[1]).length) }
```

**Comment by dschalk** [comment]
Update: The Todo list section is a React component now. The code has expanded to include rollover buttons.

**Comment by mweststrate** [comment]
Hi David,

Some answers :)

1. Thanks for pointing out the issue in the todo-mvc example! Somehow I completely broke it by using the wrong api after upgrading to 1.0, without noticing. No automated tests always bites you at some point :hammer: 

2. The loop issue is a common mistake people make when using javascrip; the `k` value is reused in all iterators of the loop because they are all reading the same closure. So you will end up with only the latest value everywhere. A simple work around is to store the variable in a variable in a fresh closure:

```javascript
for (let k=0;k<l;k+=1) {
    () => {
        const iteratorValue = k;
       // use iteratorValue instead of k from now on. 
    }()
}
```

3. I will add utilities to complete remove key/values from an object soon.

4. To simplify the splice, you can also use the `find` function, which is available in ES 7, but also on observable arrays. It takes a predicate and returns the first index that satisfies the predicate.

5. Whether a list is better or worse than an object to store that depends on many things. Not only the size of the collection, but also on whether the ordering of the the data set is important etc. (objects are unordered in javascript, although they often *seem* to be ordered). So its hard to say anything generic about that :) In performance critical applications with large datasets, and significantly more reads than writes on the data, you might even end up with a double administration of such data sets.

**Comment by dschalk** [comment]
Thank you for graciously taking the time to comment on my Todo code. I knew there was something like 'find' out there but now I know exactly what and where it is.

"for (let k=0;k<l;k+=1) " works fine, as you can see at http://transcendent.ninja . It was when I pre-defined 'k' and ran for (k=0;k<l;k+=1) that the last value of k showed up everywhere. I want to understand this better, so I am going to try "for (var k=0;k<l;k+=1)" and see if substituting 'var' for 'let' makes a difference. I don't think it will work with 'var'.

Here's something from developer.mozilla.org:
```
var list = document.getElementById("list");

for (var i = 1; i <= 5; i++) {
  var item = document.createElement("LI");
  item.appendChild(document.createTextNode("Item " + i));

  let j = i;
  item.onclick = function (ev) {
    console.log("Item " + j + " is clicked.");
  };
  list.appendChild(item);
}
The example above works as intended because the five instances of the (anonymous) inner function refer to five different instances of variable j. Note that it does not work as intended if you replace let by var or if you remove the variable j and simply use the variable i in the inner function.
```
Congratulations on publishing version '1'.  You are making a major contribution.

**Comment by dschalk** [comment]
P.S. As expected, the buttons generated by "for (var k=0;k<l;k+=1)" don't work at all. 'let' is what made the loop work without the additional boilerplate that 'var' would make necessary.

**Comment by mweststrate** [comment]
Just released 1.0.2, which introduces `mobservable.map()` can now be used to create an observable key/value maps, similar to [ES6 maps](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) (and in contrast  to objects which are expected to have more or less stable keys)
